### PR TITLE
Implementing TagsListView

### PR DESCRIPTION
**Pull request by AnnaBanani**
Implementing TagsListView

### CODE REVIEW

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        self.showMoreTagsButton.addTarget(self, action: #selector(showMoreButtonDidTap), for: .touchUpInside)
+        self.showMoreTagsButton.frame = CGRect(origin: .zero,
+                                               size: self.showMoreTagsButton.intrinsicContentSize)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func sizeThatFits(_ size: CGSize) -> CGSize {
```
Comment (Ssimboss) [inline TheMET/TheMET/TagListView/TagListView.swift:66]: This method significantly duplicates the `layoutTagView` one. Let's split out a common part of these methods. This method should receive a `width` value to calculate next outcome:
- The frames of all tag-views
- The frame of the `shoreMoreButton` (optional)
- The result layout's `height` 

For simplicity all three values could be combined in a Tuple. (Please google and read about tuples in [this arcticle](https://abhimuralidharan.medium.com/tuple-in-swift-a9ddeb314c79) )

So the API of this method would look smth like:
```
func calculateLayout(for width: CGFloat) -> (tagViewFrames: [CGRect], showMorButtonFrame: CGRect?, height: CGFloat)
```

Then use the method in both `sizeThatFits` and `layoutTagViews`. First one could just use a `height`, while a second should use all the components of the result and setup `tagViews'` frames, `showMoreButton's` frame and `totalHeight`

**Review by Ssimboss — CHANGES_REQUESTED** [review]

File: TheMET/TheMET/TagListView/TagListView.swift (hunk 1 above)
Comment (AnnaBanani) [reply TheMET/TheMET/TagListView/TagListView.swift:66]: fixed

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        let visibleTagsCount = visibleTags.count
+    throughTags: for (tagViewIndex, tagView) in tagViews.enumerated() {
+        if tagViewIndex < visibleTagsCount {
+            self.addSubview(tagView)
+            tagView.apply(radius: 15, backgroundColor: UIColor(named: "plum"), fontColor: UIColor(named: "blackberry"), font: NSLocalizedString("san_serif_font", comment: ""), fontSize: 13, buttonTitle: (self.tagViews[tagViewIndex].titleLabel?.text)!)
+            let tagFrameOrigin = self.calculateLayout(for: self.bounds.width).tagViewFrames[tagViewIndex].origin
+            tagView.frame.origin = tagFrameOrigin
+        }
+    }
+        if let showMoreButtonFrame = self.calculateLayout(for: self.bounds.width).showMoreButtonFrame,
```
Comment (Ssimboss) [inline TheMET/TheMET/TagListView/TagListView.swift:159]: You are executing `self.calculateLayout(...)` 3 times in this method. It doesn't make sense. Please execute it a single time an use the results.

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                break throughTags
+            } else {
+                let restTagsAndShowMoreButtonFrames = self.setRestTagsAndShowMoreButtonFrames(from: tagViewIndex, to: self.tagViews.count - 1, xSpace: xSpace, ySpace: ySpace, x: x, y: y)
+                result.tagViewFrames.append(contentsOf: restTagsAndShowMoreButtonFrames.tagViewFrames)
+                result.showMoreButtonFrame = restTagsAndShowMoreButtonFrames.showMoreButton
+                break throughTags
+            }
+        }
+    }
+        self.totalHeight = self.tagViews[0].bounds.height * CGFloat(lineNumber + 1) + ySpace * CGFloat(lineNumber)
```
Comment (Ssimboss) [inline TheMET/TheMET/TagListView/TagListView.swift:120]: `totalHeight` must not be set in this method. This method must only calculate a `height` for the provided width. `totalHeight` supposed to be set in `layoutTagViews` method.

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        var result: CGFloat = 0
+        for index in index1 ... index2 {
+            let tagView = self.tagViews[index]
+            result += tagView.bounds.width
+        }
+        result += xSpace * CGFloat(index2 - index1) + self.tagViews[index1 - 1].bounds.width
+        return result
+    }
+    
+    private func setRestTagsFrames(from index1: Int, to index2: Int, xSpace: CGFloat, ySpace: CGFloat, x: CGFloat, y: CGFloat) -> [CGRect] {
```
Comment (Ssimboss) [inline TheMET/TheMET/TagListView/TagListView.swift:177]: this method doesn't set tags frames anymore. Rename.

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        var xCoord: CGFloat = x
+        for index in index1 ... index2 {
+            let tagView = self.tagViews[index]
+            tagViewFrames.append(CGRect(x: xCoord, y: y, width: tagView.bounds.width, height: tagView.bounds.height))
+            xCoord += tagView.bounds.width + xSpace
+        }
+        return tagViewFrames
+    }
+    
+    private func setRestTagsAndShowMoreButtonFrames(from index1: Int, to index2: Int, xSpace: CGFloat, ySpace: CGFloat, x: CGFloat, y: CGFloat) -> (tagViewFrames: [CGRect], showMoreButton: CGRect) {
```
Comment (Ssimboss) [inline TheMET/TheMET/TagListView/TagListView.swift:188]: 1. This method doesn't set tags and showButton frames anymore. Rename.
2. The result tuple is inconsistent. You use term `tagViewFrames` for multiple frames but `showMoreButton` for a single frame. I suggest to rename the last one to `showMoreButtonFrame`

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        var result: ([CGRect], CGRect) = ([], .zero)
+        var xCoord: CGFloat = x
+        for index in index1 ... index2 {
+            let tagView = self.tagViews[index]
+            if tagView.bounds.width > self.bounds.width - xCoord - self.showMoreTagsButton.bounds.width {
+                self.showMoreTagsButton.frame.origin = CGPoint(x: xCoord, y: y)
```
Comment (Ssimboss) [inline TheMET/TheMET/TagListView/TagListView.swift:194]: This method must not set the origin of the `self.showMoreTagsButton`, it should only calculate a `showMoreButtonFrame`

**Review by Ssimboss — CHANGES_REQUESTED** [review]

File: TheMET/TheMET/TagListView/TagListView.swift [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        let estimatedHeight: CGFloat = self.calculateLayout(for: size.width).height
+        return CGSize(width: size.width, height: estimatedHeight)
+    }
+    
+    @objc
+    private func showMoreButtonDidTap() {
+        self.maxNumberOfLines = .max
+    }
+    
+    typealias frameData = (tagViewFrames: [CGRect], showMoreButtonFrame: CGRect?, height: CGFloat)
```
[the remark this answers is not in the document]
Comment (AnnaBanani) [reply TheMET/TheMET/TagListView/TagListView.swift:76]: fixed

File: TheMET/TheMET/TagListView/TagListView.swift (hunk 2 above)
Comment (AnnaBanani) [reply TheMET/TheMET/TagListView/TagListView.swift:159]: fixed

File: TheMET/TheMET/TagListView/TagListView.swift (hunk 3 above)
Comment (AnnaBanani) [reply TheMET/TheMET/TagListView/TagListView.swift:120]: fixed

File: TheMET/TheMET/TagListView/TagListView.swift (hunk 5 above)
Comment (AnnaBanani) [reply TheMET/TheMET/TagListView/TagListView.swift:188]: fixed

**Pull request merged**